### PR TITLE
[MRG] EHN add a model to fit and predict heart-rate from power

### PR DESCRIPTION
**Pull request by glemaitre**
Partially addressing #5 

Implement a scikit-learn like regressor to fit and predict heart-rate data from power.

### CODE REVIEW

**Comment by glemaitre** [comment]
@AartGoossens I modify the function which was making the estimation and prediction of the heart rate to follow the scikit-learn API. I have couple of suggestions:

* Assuming that a user pass dataframe or series, then I think it would be useful to return such type at `predict`. It can  be easily done by using the index of `X` and creating the dataframe or the series on the fly.
* We should make a sanity check and decide a policy regarding the sampling time consistency. At `fit` we should check that `X` and `y` are sampled is the same manner. At `predict` time, we need to check that we have the same sampling rate than at `fit`. This is only possible when given a datetime index with a serie or dataframe thought. Then, we need to decide whether we make a conversion or raise an error.

I also want to improve the current test. Checking strictly floating values is a bit flaky. I still don't have a clear idea thought.

Let me know if you have any comment.

**Comment by glemaitre** [comment]
I forgot to ping @sladkovm as well, sorry.

**Comment by AartGoossens** [comment]
Nice work! And good to see that you're picking up this model first because I think there are many interesting use cases for this (thinking e.g. OpenData).

- @sladkovm and me had discussions about input and return types before ([ref1](https://github.com/GoldenCheetah/sweatpy/issues/5), [ref2](https://github.com/GoldenCheetah/sweatpy/pull/14)). I'm not in favor of explicitly handling multiple input types (but Numpy and probably Scipy accept both lists/ndarrays in some cases) and am even less in favor of converting the return type to the original input type. I think the interface of the library (and therefore the function signature) should be as concise as possible as this results in the most predictable behavior for the user of the library. Maybe you can you explain how Scipy is handling this? There must some good use-cases for this.
- Good point. I have thought about this before but haven't come to a satisfactory solution. I think that for now I would be happy with just accepting 1Hz sampled data because I'm not sure what other sampling frequencies and irregular sampled data would mean for the validity of the model. Another option could be to allow specifying the sampling frequency with an optional input argument (with a default of 1).
- About testing: I really have no idea what would be the best way to do this and I agree that checking floats is flaky. [pytest.approx](https://docs.pytest.org/en/latest/reference.html#pytest-approx) might be a step in the right direction.

**Comment by glemaitre** [comment]
> but Numpy and probably Scipy accept both lists/ndarrays in some cases

`check_array` will convert list/ndarray/dataframe/series into ndarray such that we can work directly with numpy and scipy.

Regarding the output type, I really think that it would a plus to return the same type as input since we just need to reuse the index. It is coming almost for free and can avoid user to do that outside.

> I think that for now I would be happy with just accepting 1Hz sampled data because I'm not sure what other sampling frequencies and irregular sampled data would mean for the validity of the model.

I see. We could resample the data using pandas in case that we get a dataframe. In case of numpy we just assume that this is 1Hz. It should be explained in the docstring thought.

File: sksports/model/heart_rate.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    """Residuals between prediction and ground-truth."""
+    return heart_rate - _model(power, *params)
+
+
+def _model_residual_least_square(params, heart_rate, power):
+    """Least-square error of the residuals."""
+    return np.sum(_model_residual(params, heart_rate, power) ** 2)
+
+
+class HeartRateRegressor(BaseEstimator, RegressorMixin):
```
Comment (AartGoossens) [inline sksports/model/heart_rate.py:98]: Can you explain the benefit of `BaseEstimator` and `RegressorMixin`? I don't see it used anywhere in the `fit` and `predict` methods. Is there other functionality of these classes that can be useful?
Either way, I like implementing this model like a regressor class, it makes the api much more clear.

File: sksports/model/heart_rate.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    power_corrected = power + power.cumsum() * hr_drift
+    hr_lin = power_corrected * hr_slope + hr_rest
+    hr_ss = np.minimum(hr_max, hr_lin)
+
+    diff_hr = np.pad(np.diff(power_corrected), pad_width=1, mode='constant')
+    rate = np.where((diff_hr <= 0), rate_decay, rate_growth)
+
+    hr_pred = np.ones(power.shape) * hr_rest
+    for curr_idx in enumerate(hr_pred.size):
+        hr_prev = hr_pred[curr_idx] if curr_idx > 0 else hr_rest
```
Comment (AartGoossens) [inline sksports/model/heart_rate.py]: Hmm it appears that this was a bug in my sweatpy code: it always used `hr_rest`. I think you fixed that.

File: sksports/model/heart_rate.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+        Returns
+        -------
+        self
+
+        """
+        X, y = check_X_y(X, y, ensure_2d=False)
+        params_init = [self.hr_rest, self.hr_max, self.hr_slope, self.hr_drift,
+                       self.rate_growth, self.rate_decay]
+
+        if self.solver == 'leastsq':
```
Comment (AartGoossens) [inline sksports/model/heart_rate.py]: Can you explain this if-else? To me it seems the flow for `self.solver == 'leastsq'` is redundant. I.e. why can `minimize` be used for least-squares?

File: sksports/model/heart_rate.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+                 hr_slope=0.30, hr_drift=3e-5, rate_growth=24, rate_decay=30):
+        self.solver = solver
+        self.hr_rest = hr_rest
+        self.hr_max = hr_max
+        self.hr_slope = hr_slope
+        self.hr_drift = hr_drift
+        self.rate_growth = rate_growth
+        self.rate_decay = rate_decay
+
+    def fit(self, X, y):
```
Comment (AartGoossens) [inline sksports/model/heart_rate.py]: Side note: It still itches when I see Python variables starting with a capital... 😅 (but I recognize that it's accepted for these cases)

File: sksports/model/heart_rate.py [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+    .. [1] de Smet, Dimitri, et al. "Heart rate modelling as a potential
+       physical fitness assessment for runners and cyclists." Workshop at ECML
+       & PKDD. 2016.
+
+    """
+
+    def __init__(self, solver='Nelder-Mead', hr_rest=75, hr_max=200,
+                 hr_slope=0.30, hr_drift=3e-5, rate_growth=24, rate_decay=30):
+        self.solver = solver
+        self.hr_rest = hr_rest
```
Comment (AartGoossens) [inline sksports/model/heart_rate.py]: Minor: I'd like to propose renaming `hr_rest` to `hr_start` or `hr_0` because this model parameter is not related to the actual resting heart rate of the athlete and therefore might cause confusion.

File: sksports/model/heart_rate.py (hunk 4 above)
Comment (glemaitre) [reply sksports/model/heart_rate.py]: Yep, this is kinda of imposed by scikit-learn estimators. `power` and `heart_rate` would be much better.

File: sksports/model/heart_rate.py (hunk 3 above)
Comment (glemaitre) [reply sksports/model/heart_rate.py]: `'leastqs` is actually using Levenverg-Marquardt algorithm. I was just playing with the different way of minimizing the problem. I am not really sure what is the best solution there. We can of course fix the solver as well.

NB: Levenvenberg-Maquardt will use  the gradient will an algorithm as the simplex (Nelder-Mead) will not use it. So depending of the optimization problem and the convexity it would be more appropriate to use one or another.

File: sksports/model/heart_rate.py (hunk 1 above)
Comment (glemaitre) [reply sksports/model/heart_rate.py:98]: basically, it allows to use the tool as pipeline, grid-search, and cross-validation from scikit-learn.
The mixin just define a `score` function which is the R^2 score while the `BaseEstimator` provides `get_params` and `set_params` to deal with the parameters.

Comment (glemaitre) [reply sksports/model/heart_rate.py:98]: > Either way, I like implementing this model like a regressor class, it makes the api much more clear.

We can also the model public to the user. We just need a good name.

File: sksports/model/heart_rate.py (hunk 5 above)
Comment (glemaitre) [reply sksports/model/heart_rate.py]: good point, I was confused :) `hr_start` seems a good idea,

**Comment by glemaitre** [comment]
On a simple example (which will be in the tests) it seems that Nelder-Mead is the only solver leading to the right parameter. I think that we could simplify and keep the solver that you propose at the beginning.

File: sksports/model/heart_rate.py [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+
+    def _check_inputs(self, X, y=None):
+        """Validate the inputs.
+
+        Check if the inputs are Pandas inputs. Resample with a sampling-rate of
+        1 Hertz.
+
+        """
+        if y is None:
+            is_df = True if hasattr(X, 'loc') else False
```
Comment (AartGoossens) [inline sksports/model/heart_rate.py:170]: Although this is valid in most cases, why not just go for `isinstance(X, pd.DataFrame)`?

File: sksports/model/heart_rate.py [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+            X_ = X.resample('1s').mean() if is_df else X
```
Comment (AartGoossens) [inline sksports/model/heart_rate.py]: I think this raises a `TypeError` for a df with a non datetime-like index. Should we catch that here or add a check for a datetime-like index?

File: sksports/model/heart_rate.py [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+            X_ = check_array(X_)
+            return X_.ravel(), is_df
```
Comment (AartGoossens) [inline sksports/model/heart_rate.py]: You still need to return `y` here as the second return argument right? Else line 193 will raise a `ValueError` (unpack error).

File: sksports/model/heart_rate.py [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+
+        Returns
+        -------
+        y : ndarray, shape (n_samples,)
+            The heart-rate predictions.
+
+        """
+        check_is_fitted(self, ['hr_start_', 'hr_max_', 'hr_slope_',
+                               'hr_drift_', 'rate_growth_', 'rate_decay_'])
+        power, is_df = self._check_inputs(X)
```
Comment (AartGoossens) [inline sksports/model/heart_rate.py]: I think `self._check_inputs` returns 3 items

File: sksports/model/tests/test_heart_rate.py [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+# License: MIT
+
+import pytest
+import numpy as np
+import pandas as pd
+
+from sksports.model import exp_heart_rate_model
+from sksports.model import HeartRateRegressor
+
+POWER = np.array([100] * 10 + [240] * 100 + [160] * 100)
```
Comment (AartGoossens) [inline sksports/model/tests/test_heart_rate.py]: You can consider moving these constants to a pytest fixture. It's a matter of preference but I think these fixtures are really nice to work with and also well reusable in other tests. But I'm fine with doing improvements like this later.

File: sksports/model/tests/test_heart_rate.py [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+      pd.Series(
+          HEART_RATE,
+          index=pd.date_range('1/1/2011', periods=POWER.size, freq='s')),
+      np.ndarray)]
+)
+def test_heart_rate_regressor(power, heart_rate, output_type):
+    reg = HeartRateRegressor()
+    reg.fit(power, heart_rate)
+
+    assert reg.hr_start_ == pytest.approx(100, 1.0)
```
Comment (AartGoossens) [inline sksports/model/tests/test_heart_rate.py]: More of a question than a comment: what's best practice for models like this for deciding how and which model parameters to test? Do you have any examples from scikit-learn?

**Comment by AartGoossens** [comment]
> On a simple example (which will be in the tests) it seems that Nelder-Mead is the only solver leading to the right parameter. I think that we could simplify and keep the solver that you propose at the beginning.

This is really interesting. What kind of results did you get? I expected other solvers to work at least reasonably well. I'm already looking forward to using OpenData to play around with different solvers... (it's time for me to get some more work into that the OpenData tooling)

File: sksports/model/heart_rate.py (hunk 6 above)
Comment (glemaitre) [reply sksports/model/heart_rate.py:170]: It was to ducktype. It usually allows to not import the library (which we do anyway) and accept data structure which are inheriting from it.

File: sksports/model/heart_rate.py (hunk 7 above)
Comment (glemaitre) [reply sksports/model/heart_rate.py]: Yes, dataframe is not enough, we should check the index.

File: sksports/model/heart_rate.py (hunk 8 above)
Comment (glemaitre) [reply sksports/model/heart_rate.py]: Since it is confusing, I should always return `X_.ravel(), y, is_df` and ignore `y` when it is None.
Right now, I am returning different thing depending if it is `fit` or `predict`.

File: sksports/model/heart_rate.py (hunk 9 above)
Comment (glemaitre) [reply sksports/model/heart_rate.py]: only 2 at predict

File: sksports/model/tests/test_heart_rate.py (hunk 10 above)
Comment (glemaitre) [reply sksports/model/tests/test_heart_rate.py]: Good point. I was playing with the tests for the moment :)

File: sksports/model/tests/test_heart_rate.py (hunk 11 above)
Comment (glemaitre) [reply sksports/model/tests/test_heart_rate.py]: Usually we run a classifier or a regressor on a known dataset and check the score. It is non-trivial to actually test parameters. Here I think that we can make it on those 2 parameters with a large enough approximation to be sure that the optimization (or any change in the future) does not lead to a complete non-sense.

Basically, I think that we could make something like:

```python
assert reg.score(X) > 0.9
```

to be sure that we are fitting the data properly.

**Comment by glemaitre** [comment]
> What kind of results did you get?

solver

File: sksports/model/heart_rate.py (hunk 8 above)
Comment (AartGoossens) [reply sksports/model/heart_rate.py]: Ah now I get it. That could work indeed but I think it's less confusing if a method always returns the same number of arguments.

File: sksports/model/tests/test_heart_rate.py [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+    (pd.DataFrame, pd.timedelta_range(start='0 day', periods=power_arr().size,
+                                      freq='s'))])
+def power(request, power_arr):
+    power = request.param[0](power_arr.reshape(-1, 1))
+    if request.param[1] is not None:
+        power.index = request.param[1]
+    return power
+
+
+@pytest.fixture(params=[
```
Comment (AartGoossens) [inline sksports/model/tests/test_heart_rate.py:34]: This looks interesting! If I'm correct you're not parametrizing the test but the fixture?...

Comment (glemaitre) [reply sksports/model/tests/test_heart_rate.py:34]: Yes.I would think that fixture are good for data (parametrizing allows to build different supported type. Then parametrizing the test is useful for passing some parameters to the function or something specific to the test. This is just my thumb rule and not sure this is the best ;)

**Comment by glemaitre** [comment]
Links to the documentation:

- User guide: https://12-134096781-gh.circle-artifacts.com/0/doc/inference.html#heart-rate-inference
- Example: https://12-134096781-gh.circle-artifacts.com/0/doc/auto_examples/model/plot_heart_rate.html#sphx-glr-auto-examples-model-plot-heart-rate-py

File: examples/model/plot_heart_rate.py [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+# of the model.
+
+from sksports.model import exp_heart_rate_model
+
+# The function `exp_heart_rate_model` required numpy array
+heart_rate = exp_heart_rate_model(data['power'].values, 75, 200, 0.4, 3e-5,
+                                  15, 40)
+data['heart-rate'] = heart_rate
+data.plot(legend=True)
+plt.title('Simulated power of a cyclist and inferred heart-rate.')
```
Comment (glemaitre) [inline examples/model/plot_heart_rate.py]: remove final dot

File: examples/model/plot_heart_rate.py [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+print('The fitted parameters of the regressor are: \n {}'
+      .format({'hr_start_': reg.hr_start_, 'hr_max_': reg.hr_max_,
+               'hr_slope_': reg.hr_slope_, 'hr_drift_': reg.hr_drift_,
+               'rate_decay_': reg.rate_decay_,
+               'rate_growth_': reg.rate_growth_}))
+
+power.plot(legend=True)
+heart_rate.plot(legend=True)
+heart_rate_initial.plot(legend=True)
+heart_rate_pred.plot(legend=True)
```
Comment (glemaitre) [inline examples/model/plot_heart_rate.py:89]: add title

**Comment by glemaitre** [comment]
@AartGoossens I was checking the results of the optimization using Nelder-Mead and it can lead to fitted parameters without any mining even if the fitting is good (e.g. `hr_max` could be 300). Somehow, I was thinking to make a minimization with bounds but L-BFGS is really not stable.

Somehow I found that the best for the moment is to use Nelder-Mead at first and then refine with L-BFGS using bounds. 

One of the drawback is that we make the fitting slower (i.e. we are minimizing twice). I also not sure 100% if combining both methods will be always stable or if L-BFGS can give erroneous results as when we use it alone.

**Comment by glemaitre** [comment]
Also `HeartRateRegressor` is maybe not the best name. If we go some other methods to do the same job then we will have some issue with naming.

**Comment by AartGoossens** [comment]
That's very interesting indeed. Is combining optimizations common practice? I haven't heard of it before.

If this is only about a better fit for the `hr_max` I want to mention this: I do not expect that the `hr_max` model parameter is really interesting for 2 reasons:

1. I expect it to properly minimize only for activities where there's enough high power data (i.e. where `hr_lin > hr_max`)
2. Heart rate and power are only approximately linear correlated for intensities where the almost all work is done aerobically (I think this is also mentioned in the article of De Smet). As a result the current model will overestimate heart rate for higher intensities and of course this will result in a worse model fit.

In conclusion: in order to properly fit/minimize the hr_max we need activities with high power (1) but we also know that for these activities the model fit is not that good (2)...

Either way, if the fit is improved by combining these optimizations it's worth it of course.
By the way, in my opinion the performance hit is not an issue.
